### DISCUSSION TITLE
Textures will turn "Black" in AR mode <model-viewer> using KTX 2.0 compression METHOD 3: ETC1S

### DISCUSSION THREAD

**Discussion by andygomez3dart** [google/model-viewer#2348] [Q&A]
Textures will turn "Black"  in AR mode <model-viewer> using KTX 2.0 compression METHOD 3: ETC1S

In my example I used the following textures:
Diffuse : JPG file,  2048x2048 pixels
Normal: PNG file, 1024x1024
glb size before compression : 10.9 MB
glb size after compression : 2.17 MB

Everything works great and fast in the viewport <model-viewer> when using a glb file with the METHOD 3: ETC1S.
The problem comes when I start the AR (augmented reality) mode and the 3D model turns "Black".

I wonder whether there is a problem with the way I applied the textures or it's just an incompatibility between <model-viewer> & KTX 2.0 compression at the moment.

Please find attached  the screenshots below:

Any idea about what it's going on?

![VID-20210502-WA0001](https://user-images.githubusercontent.com/79045928/116808898-2d805900-ab33-11eb-983a-da0894071a94.gif)
![VID-20210502-WA0002](https://user-images.githubusercontent.com/79045928/116808904-3113e000-ab33-11eb-92d5-485db8859c0a.gif)

**Comment by elalish** [comment]
From the hand icon, this appears to be a Scene Viewer bug, @tpsiaki. Is WebXR mode working alright?

**Reply by andygomez3dart** [reply]
I am using the `<model-viewer>` version 1.6.0 if that's helpful.

**Reply by elalish** [reply]
Can you post the model or a repro link so we can figure out what's going on?

**Reply by andygomez3dart** [reply]
Hi @elalish I thought you were right. About the HTTP but I still see the textures are not displaying properly. Still I can see "black" from Android device You can see it [here](https://www.ag3dart.com/ct-scan-3d-models-female-skeleton)  thanks very much once again

**Reply by elalish** [reply]
I'm glad WebXR is working, but the black textures are still a problem in Scene Viewer. Can you post the model here so we can get that fixed?

**Reply by elalish** [reply]
I see some interesting error and warnings in your console. The most relevant to black textures looks to be `THREE.WebGLRenderer: Texture is not power of two. Texture.minFilter should be set to THREE.NearestFilter or THREE.LinearFilter.`

We've noticed some glTF/KTX2 exporters that are making errors like this, for instance by not including mipmaps in the KTX2 file. They must be included because they cannot be generated at runtime for compressed textures.

**Reply by elalish** [reply]
Ah, I think I figured out why you're not getting WebXR mode even with HTTPS; we really need to improve our error/warning reporting. I believe you're in an iframe? You'll need to specifically allow webxr to run in that context or it will fail: https://medium.com/samsung-internet-dev/using-webxr-in-html-iframes-629c248ff96d

**Reply by andygomez3dart** [reply]
> I see some interesting error and warnings in your console. The most relevant to black textures looks to be `THREE.WebGLRenderer: Texture is not power of two. Texture.minFilter should be set to THREE.NearestFilter or THREE.LinearFilter.`
> 
I am not familiar with "Three" but I am surprised  the texture is not power of two cause I used a 2048x2048 and when I converted to KTX via cmd didn't give errors about  textures regarding power of two, It's very wierd.
![Diffuse Texture](https://user-images.githubusercontent.com/79045928/117342748-7ae62880-ae9b-11eb-9287-3c8f8740c30f.png)

**Reply by andygomez3dart** [reply]
> Ah, I think I figured out why you're not getting WebXR mode even with HTTPS; we really need to improve our error/warning reporting. I believe you're in an iframe? You'll need to specifically allow webxr to run in that context or it will fail: https://medium.com/samsung-internet-dev/using-webxr-in-html-iframes-629c248ff96d

Unfortunately I need to use an iframe to to able to use HTML & JS in my website...
When you say "you'll need to specifically allow webxr " Do you mean to add "webxr" as an attribute sort of like "allowfullscreen" within the iframe ?  Sorry I just recently started my journey with  web development 😊 
@elalish Many thanks for such comprehensive analysis

**Reply by elalish** [reply]
Yeah, I would take the exact text with a grain of salt, especially considering how new KTX2 support is. Check that your KTX2 file has mipmaps included; it's not technically required, but you will not get good results if you don't.

**Reply by elalish** [reply]
Please read the link; it says exactly what you have to add.

**Reply by andygomez3dart** [reply]
@elalish  I did and it's very simple to use. I can see that works in the example you provided, unfortunatley for me  I don't have access to ifames within the top level HTML of my website ,😢,  any chance to implement what you suggest in a  lower level HTML where I have my code for the  `model-viewer` web component ? I'm really frustrated with my web hosting provider...

**Reply by elalish** [reply]
Unfortunately this is a browser security boundary, so I don't think you can get around it. @klausw in case he has any more insight on the WebXR side of things.

**Reply by klausw** [reply]
As far as I know there's no way for the inner iframe to override the outer page's security policy, and this is by design. You either need to get the outer page to allow `xr-spatial-tracking`, or reload your page to run at the toplevel instead of in an iframe, for example by setting `window.top.location`.

**Reply by andygomez3dart** [reply]
Many thanks for all your support guys, I guess for now I will have to keep the normal glb file as it is before compression, 
I really wanted to find out If I could use a compressed glb file using KTX 2.0  when entering in AR mode  webxr, scene-viewer etc.
Do you think I can achieve this by using Wordpress as a webhosting ?

**Reply by tom-deadline** [reply]
@elalish Could you suggest what ktx2 compressor to use?

**Reply by elalish** [reply]
This is all new enough I'm really not sure. If you find some good info, please start a discussion to spread the word. I hope to eventually include this as a feature in our editor.

**Reply by tom-deadline** [reply]
@elalish I get these errors in editor: 
`Cannot validate an extension as it is not supported by the validator: 'KHR_draco_mesh_compression'.`
`Invalid value 'image/ktx2'. Valid values are ('image/jpeg', 'image/png').`

is it editor's validator issue or model-viewer doesn't support draco and ktx2 anymore?

**Reply by elalish** [reply]
No, sorry, the validator just needs an update. If those are the only errors, the file should be fine.
